### ISSUE TITLE
Deploy Source to org command takes several minutes to complete

### ISSUE THREAD

**Issue by WuggyX2** [forcedotcom/salesforcedx-vscode#4865]
### Summary

We have noticed that when using the "deploy source to org" command from the command palette, the notification that the action in running does not close even though the deployment has already completed in the backgroud. This give a wrong impression that the deployment takes several several minutes to complete. Some times the deployed to source message gets printed twice even though the command was run only once.

### Steps To Reproduce:

1. Create a new project or take an existing one and connect it to an org of your choosing.
1. Do a change to a lightning web component
1. Run the "deploy source to org" command

### Expected result

Success message for the deployment and the list of deployed components should appear in a second or two

### Actual result

Success message takes several minutes to appear. Sometimes the command in progress notificaiton keeps running even though list of deployed components has already been printed.

### Additional information
![kuva](https://github.com/forcedotcom/salesforcedx-vscode/assets/51646225/662e8d4e-8dda-4c29-8cbc-507170164aa3)

Here's an example when deployed to source gets printed twitch even though the command is run only once
![kuva](https://github.com/forcedotcom/salesforcedx-vscode/assets/51646225/f9265c54-8174-4165-95f6-ca28b67d14b9)

**Salesforce Extension Version in VS Code**:
v57.13.1

**SFDX CLI Version**:
@salesforce/cli/1.78.0 darwin-x64 node-v18.15.0

**OS and version**:
MacOS Ventura 13.3.1

**Comment by randi274** [comment]
@WuggyX2 I've noticed this recently as well - if you cancel the first deploy and then run the command again, do you also find it's much faster on the second run?

**Comment by WuggyX2** [comment]
@randi274 Unfortunately canceling the first deploy does not seem to increase the speed of the second deploy. However, removing the `org-metadata-rest-deploy`config variable for the old deploy/retrieve commands seems to have helped for now at least for me. That variable also breaks the default org switcher in the status bar.

**Comment by mnunezdm** [comment]
In my case, if i go to the environment I find that deployment is done like normally but vscode keeps polling

**Comment by randi274** [comment]
@WuggyX2 or others - what type of orgs are you working against?

**Comment by WuggyX2** [comment]
@randi274 Me and my team are working against developer sandboxes.

**Comment by JavierDAP** [comment]
I don't know if it is related but I noticed that now .sf\orgs\ is getting populated with the source tracking. With each deploy that I do, the folders gets bigger and bigger and makes things slower.
 So if I start a new project  the first deploys are fast but little by little as more files are created (automatically) in .sf\orgs\ deploys get longer and longer. (I wish I could disable source tracking)

**Comment by randi274** [comment]
@JavierDAP this is exactly the change we're thinking might be the cause here. Are you working with a source-tracked org (like a scratch org or source-tracked sandbox) or also in a non-source tracked org (like a developer org)?

**Comment by JavierDAP** [comment]
Yes in the Dev Hub it is activated Source Tracking in Developer and Developer Pro Sandboxes

**Comment by randi274** [comment]
Thanks all for the feedback so far! We recently fixed-up source tracking for source-tracked sandboxes, and we believe this is the functionality you're now encountering. Previously, source tracking wasn't being fixed up for source-tracked sandboxes at all (so you couldn't run commands like `View All Changes` to compare the code you're working with locally vs in the org you're connected to), but now we're finding there might be a little more cleanup work we need to do to make this fixing up process more efficient.

A few follow up questions for us to confirm what's happening:
0 - Are your sandboxes all source tracked? @JavierDAP sounds like that's the case for you.
1 - Are deploys successful when you expect them to be? Or are some not completing at all?
2 - If they're all successful, is the notification lingering even after you've confirmed deployment in your org? @mnunezdm that sounds like this is your experience.
3 - When did you first notice the issue?
4 - And finally - have you worked with the new source tracking functionality? Is this something you want to be able to use during your development?

**Comment by JavierDAP** [comment]
0 - I'm not sure how to see if an specific org is source tracked. I guess all are, because some of them are new or were recently refreshed.
1 - Some not get completed at all, but normally only takes a couple minutes.
2 - Seeing "Deploy Status" normally it gets deployed instantly but the notification lingers a lot.
3 - This Wednesday (17/05)
4 - First time this new system. Could be useful, specially the first time that you create a local project to download all the data from the org instead of using a package.xml. Using git in visual studio code is a little bit annoying because create a lot of files so we are going to add .sf/orgs to .gitignore. 

As a side note, I don't know if source tracking could enable to retrieve the totality of a profile metadata with a simple retrieve. Because right now, it is necessary to have a package.xml with all metadata and retrieve that file to get all settings on the profile and managing this in git is very tedious.

**Comment by mnunezdm** [comment]
hello @randi274, in my case source tracking is disabled for sandboxes (setting in production)

Pretty sure this has something to do with source tracking, because after a deployment the source control sidebar shows is loading

The deployment in the org finishes almost instantly, the vscode takes like 1-2 minutes to finish

While deploying vscode goes unresponsive and the cpu usage goes up, navigating to classes takes a while. When the deployment finishes in vscode everything goes back to normal

I have included a video showing the behavior when deploying with the extensions https://youtu.be/IeYm5vBQ9VM

Doing the deployment with the command line is a good workaround

Also for me, I think it was with the latests update of the vscode extensions (v57.14.1) currently using latest sfdx on stable-rc 7.202.7-c8df4cc

**Comment by JayverLance** [comment]
Is this issue now leading to the issue reported in https://github.com/forcedotcom/salesforcedx-vscode/issues/4873?  I seem to have encountered the Source Tracking last week and then issue with Deployment followed.

**Comment by JavierDAP** [comment]
@JayverLance for me the issue described in #4873 happens when I delete the files created by source tracking.

**Comment by JayverLance** [comment]
I've just resolved this issue on my VS Code by reverting Salesforce Extension Pack to v57.10.2. I've also installed the older versions of Apex, Aura Components, Apex Replay Debugger, Salesforce CLI Integration, Apex Interactive Debugger, Lightning Web Components, SOQL, and Visualforce to the versions from 1 month ago. I only started having this issue last week when my VS Code automatically updated the Extensions to an update 1 week ago.

**Comment by forcetrekker** [comment]
Is the deployment slow because the VScode's context menu command is using deprecated commands?

**Comment by randi274** [comment]
As an immediate workaround for this issue, we recommending downgrading to v57.10.2 for all of the extensions. We believe this slowness is a result of fixing up the source tracking in a given project, and will need to do some work to see what the best approach is to resolve that latency.

One approach we are considering would be to put the source tracking behind a custom setting so that if you find it's not providing you value during your workflow, it doesn't have to slow you down either. If any of our intrepid users would like to try out that approach for us, we'd appreciate confirmation that it speeds up your development again! To this change, you can download that version of our VS Code Extensions [here](https://github.com/forcedotcom/salesforcedx-vscode/actions/runs/5028729765), and then in the extensions tab of VS Code select "Install from VSIX" for each of the extensions.
<img width="672" alt="Screenshot 2023-05-23 at 3 44 41 PM" src="https://github.com/forcedotcom/salesforcedx-vscode/assets/9795193/c251d1e1-c5b3-42a0-9d6a-6a7850cd9861">

**Comment by mnunezdm** [comment]
I will give it a try, where do we have to configure the "custom setting" @randi274 ?

**Comment by JavierDAP** [comment]
Same, I have it already installed but I don't see where it is

**Comment by randi274** [comment]
Thanks for jumping on it @mnunezdm and @JavierDAP! There's no setting in that version, just the exposed code that we would put behind the setting, so we'd like to know if that fix addresses the latency issue. If the performance is better with this version, then we can make a plan for adding the setting in our published extensions.

**Comment by mnunezdm** [comment]
seems to be working for me ! @randi274

**Comment by JavierDAP** [comment]
I would also say that works as before. It still retrieve source tracking data (at least in my case)

**Comment by randi274** [comment]
Thanks all for your patience so far. We've started along the path to getting a fix out to get everyone's deployments back to the state they were in before, but we've also noticed as @JavierDAP did that we're not fully there yet. This is something we are still actively working to fix.

As another workaround, if you or your admin are able to **turn off source tracking**, that is also a quick fix option (thanks to @IttesITNick for the reminder for this on another issue!):
- Navigate to setup > Dev Hub
- Disable the setting "Enable Source Tracking in Developer and Developer Pro Sandboxes"
- Refresh your sandbox and authorize VS Code again with the new sandbox

**Comment by mnunezdm** [comment]
Hello, In our org, Source tracking is disabled and as far as I know it haven't been enabled ever

**Comment by justin-lyon** [comment]
> 

Thanks. This is the best work around for me right now.

**Comment by JimR-FAN** [comment]
I started to write a lengthy "me too" explanation with more details, but it seems like it's fixed with v58.0.1.

Reverting back to  v57.7.0 is previously what worked for me, but that's because I was having a different issue that wasn't fixed until v57.15.0, which, incidentally, is when I first noticed THIS issue.   

Well, it seems like it's fixed with v58.0.1!   Nice job and thank you!

**Comment by randi274** [comment]
Update on this! We've introduced a new temporary setting as we continue to work on source tracking improvements with v58.0.1 to Enable Source Tracking. We have set this to be automatically enabled by default, as we recommend working with source tracking if you find it helpful, but we don't want this feature to continue interrupting your workflow. If you find you're still facing issues, try turning that setting to off, restarting VS Code, and check if that helps your deployment performance. You may need to clear out source tracking for a project (and beware of #4873), or start a new project to get back to a clean slate.

We'll continue to keep this one open for another release or two, so let us know if you continue to experience issues as well. Thanks all!
![Screenshot 2023-06-02 at 11 16 25 AM](https://github.com/forcedotcom/salesforcedx-vscode/assets/9795193/27752385-14e9-46b4-9d95-ddcb337b1c6d)

FYI @RitamAgrawal @klewis-sfdc @AnanyaJha

**Comment by forcetrekker** [comment]
In case you do not want to downgrade, disable [salesforce's deploy on save](https://developer.salesforce.com/tools/vscode/en/user-guide/deploy-on-save) and use CLI directly, manually running deploy command everytime is little tedious.

To automate this, I like to use [Save and Run](https://marketplace.visualstudio.com/items?itemName=wk-j.save-and-run) vscode extension, using which I configure custom deploy on save command with below setting in `.vscode/settings.json` file on my workspace:

```json
{
    "saveAndRun": {
        "commands": [
            {
                "match": "force-app/*",
                "isAsync": true,
                "cmd": "time sf project deploy start --source-dir '${file}' --ignore-conflicts"
            }
        ]
    }
}
```

Note: I run SF retrieve before working on a file, hence I ignore conflicts with `--ignore-conflicts` flag, discard it if you want to see conflicts with save.

**Comment by IvanSudakov** [comment]
@randi274 Unfortunately, disabling the Source Tracking in v58.0.1 or reverting Salesforce Extension Pack to v57.10.2 did not really fix the issues. Initially, the deployment could take up to 30 minutes, now it goes faster but still about 5 minutes. Any suggestions?  Instance | CS100, dev sandbox.

**Comment by dlnguyenn1** [comment]
> @randi274 Unfortunately, disabling the Source Tracking in v58.0.1 or reverting Salesforce Extension Pack to v57.10.2 did not really fix the issues. Initially, the deployment could take up to 30 minutes, now it goes faster but still about 5 minutes. Any suggestions? Instance | CS100, dev sandbox.

I made sure to also revert all other extensions such as Apex, the Debugger, Lightning Web Components, etc.  Just reverting the extension pack did not fix anything.  Try this as well.

**Comment by IvanSudakov** [comment]
> > @randi274 Unfortunately, disabling the Source Tracking in v58.0.1 or reverting Salesforce Extension Pack to v57.10.2 did not really fix the issues. Initially, the deployment could take up to 30 minutes, now it goes faster but still about 5 minutes. Any suggestions? Instance | CS100, dev sandbox.
> 
> I made sure to also revert all other extensions such as Apex, the Debugger, Lightning Web Components, etc. Just reverting the extension pack did not fix anything. Try this as well.

@dlnguyenn1 Thank you for the suggestion; turns out the long deployment time was not caused by CLI but by the post-deploy compilation ("Compile On Deploy" in Apex Settings). Turning this feature off bring back the deployment to about 10 seconds.

**Comment by AnanyaJha** [comment]
Hi @IvanSudakov thanks for sharing your fix! Can you share a screenshot of the setting that helped you with reducing the deployment times? And were you able to see the faster deployment times on the fully reverted version of the Extension Pack or the latest version?

@dlnguyenn1 thanks for sharing your fix as well! Did you end up updating any of the Settings, similar to what Ivan mentioned above?

**Comment by rodrigoreyes79** [comment]
Hi all. I am experiencing 10-15 min source deploy times since yesterday. "Compile On Deploy" is turned off as you can see in the attached image: 

![image](https://github.com/forcedotcom/salesforcedx-vscode/assets/3924426/fec9198a-02d0-4a07-a967-87c9fe939959)

This started happening about 2 days ago for me and I already created a new scratch org because I thought it might be related to the specific org I was working on but seems to still happen in the new org.

Anyhow, another workaround that might help other poor souls: on VS Code, set this settings like this:

![image](https://github.com/forcedotcom/salesforcedx-vscode/assets/3924426/fde480d5-8472-47c7-89e2-41f1d29f8fc3)

That should allow you to use deploy instead of push (no more long delays) and now you will push on save to avoid keeping track of what changed in your head (or somewhere else).

**Comment by RitamAgrawal** [comment]
Mentioning https://github.com/forcedotcom/salesforcedx-vscode/issues/4926 here and closing 4926 as it is the same problem and workaround.

**Comment by markgarg** [comment]
Still happening for me, and downgrading to `v58.0.1` as @JimR-FAN has mentioned above works 👍🏽

**Comment by CristiCanizales** [comment]
Hello everyone! I just wanted to verify that this is still an issue for you when running the last version of the extensions (v58.14.1)

**Comment by jaapvandermolen** [comment]
@CristiCanizales I just had a simple flow take 2m30s to deploy, so I'd say the issue has not been resolved. I upgraded to v58.14.1 specifically to test.

**Comment by jnadeausf** [comment]
Seeing same with retrieving source. It almost immediately updates the local source but the "Running SFDX:..." window stays for a minute or two.

**Comment by CristiCanizales** [comment]
@jaapvandermolen and @jnadeausf, thanks for your response. Could you tell me the OS you're running on?

**Comment by jnadeausf** [comment]
macOS Ventura 13.5

**Comment by zanon1101** [comment]
Same issue on macOS Monterey 12.5.1. I've downgraded VSCode Salesforce CLI extension to v58.11.0 and it seems to be much better

**Comment by gabo-andrade** [comment]
Tried several times with several versions. Downgrading Salesforce CLI extension solves the issue for a while. But after a few deploys the deployments become slow again

**Comment by daphne-sfdc** [comment]
Hello everyone,
We're investigating this issue and found that slower wifi speeds tend to cause longer wait times for deployment.  Can you do a speed test and send us a screenshot of the result?

**Comment by jaapvandermolen** [comment]
![image](https://github.com/forcedotcom/salesforcedx-vscode/assets/121507614/3cf87bf9-03d6-4571-8188-6bd06780fdee)

**Comment by nihi-lb-dk** [comment]
This is still an issue. It is not feasablie to use this extension during development with this state.

**Comment by gabo-andrade** [comment]
Yeah, don't think the wifi speed is related at all. 
To me, the first deploy after launching VS Code finish pretty fast, and after that each deployment takes more time. After few deploys it becomes so unbearable that I prefer to restart VS Code to deploy again. How bad the issue becomes that it really take less time to restart VS Code and deploy again.

![image](https://github.com/forcedotcom/salesforcedx-vscode/assets/1302875/b29e1cea-5eb4-4e12-8a5c-32528b67aee8)

**Comment by gabo-andrade** [comment]
`Apex Setting: Perform Synchronous Compile on Deploy` was already disabled in my org. Also I'm working with scratch orgs and this happen in all orgs.

Now disabling `Enable Source Tracking For Deploy And Retrieve` is working, deploys are all taking more or less the same time. At the moment I have only tested deploying LWC but everything goes great

**Comment by daphne-sfdc** [comment]
👋 Hello everyone!  We found another workaround that has solved the slow deployment problem for some users.

Please try going into the Explorer view in the sidebar and deleting the .sfdx and .sf folders and then running "Developer: Reload Window" from the Command Palette.  This will clean up the configs on your project and reset the authorization.  Then re-authorize to your org and try deploying again.

**Comment by kenny-buckles** [comment]
I have done both recommendations listed above 1) disable vscode setting for source tracking, and 2) deleting the .sfdx and .sf folders and then running "Developer: Reload Window" from the Command Palette. I am still having extremely slow deploy and apex test run times. Are there any further developments on this?

**Comment by ImJohnMDaniel** [comment]
curiously watching progress 😎

**Comment by mateusz-domagalski** [comment]
Great to know that it's being actively worked upon. I appreciate it.
The only workaround that has worked for me was to downgrade the extension to v57.10.2.

**Comment by CristiCanizales** [comment]
Hi @mateusz-domagalski I'm just curious, the setting `Salesforcedx-vscode-core > Experimental: Enable Source Tracking For Deploy And Retrieve` (and reloading vscode window after) didn't work for you either?

**Comment by gonardfreeman** [comment]
Nothing helped but I see this error in console (for both 57.10.2 and current)
![57 10 2](https://github.com/forcedotcom/salesforcedx-vscode/assets/16113714/30cbf91c-0840-4eaf-af2c-562df4b4bac9)
Mb it will help

**Comment by AnanyaJha** [comment]
Hi @gonardfreeman @ImJohnMDaniel @mateusz-domagalski ! We released an update last week that changed a few things under the hood to help speed up deployments - could you all take a look and let us know if you're seeing any impact to your deployment times? Thank you!

**Comment by DoganCeliktemel** [comment]
Hello, I am facing the above mentioned issue, too. I tried downgrading the VSCode extensions (to v57.10.2) and also disabled the Enable Source Tracking For Deploy And Retrieve setting in the VSCode preferences. However, my deploy source to org execution times are still very tiring (> 5 min for a single Apex class). Is there anything else you can recommend me to do?

**Comment by tadahh** [comment]
We as a development team are also facing the same issues. I have personally uninstalled the cli and all extensions and reinstalled node, npm, cli, and the vscode extension pack to latest versions. We are seeing conflict detection sitting stale for 20+ mins and every once in a while I will receive the below error


<img width="1185" alt="Screenshot 2023-10-26 at 12 59 12 PM" src="https://github.com/forcedotcom/salesforcedx-vscode/assets/5883817/e43e063a-3349-49ba-9860-75f47652cf14">

**Comment by AnanyaJha** [comment]
Hi all, thanks for sharing the issue you're running into! We'll have a member of our team take a closer look at what might be going on cc @gbockus-sf

**Comment by kenny-buckles** [comment]
I also get the same error @tadahh is getting. Not just when I use vscode for running tests/deployment, but even within Salesforce and other tools hitting various Salesforce API's.

**Comment by CristiCanizales** [comment]
Hi @DoganCeliktemel, did you reload vscode after disabling source tracking?

**Comment by CristiCanizales** [comment]
> Hello! Just wanted to know if you all have the `Apex Setting: Perform Synchronous Compile on Deploy` (which you can find **in your org settings**, in setup) enabled, and if so try:
> 
> * disable setting
> * go to vscode and reauth to your org
> * reload vscode
> 
> and see if it helps
> 
> Also, we want to know if the slow deployment happens to all your orgs or if it's ok with a new one
> 
> We're asking all these q's because disabling the setting or working with a new org has fixed it for some users. And of course disabling our vscode setting: `Salesforcedx-vscode-core > Experimental: Enable Source Tracking For Deploy And Retrieve` (and reloading vscode window after)
> 
> Thanks again for your patience and sorry for all the inconveniences this has caused

Hi @tadahh, @kenny-buckles, have you tried these workarounds?

**Comment by tadahh** [comment]
> > Hello! Just wanted to know if you all have the `Apex Setting: Perform Synchronous Compile on Deploy` (which you can find **in your org settings**, in setup) enabled, and if so try:
> > 
> > * disable setting
> > * go to vscode and reauth to your org
> > * reload vscode
> > 
> > and see if it helps
> > Also, we want to know if the slow deployment happens to all your orgs or if it's ok with a new one
> > We're asking all these q's because disabling the setting or working with a new org has fixed it for some users. And of course disabling our vscode setting: `Salesforcedx-vscode-core > Experimental: Enable Source Tracking For Deploy And Retrieve` (and reloading vscode window after)
> > Thanks again for your patience and sorry for all the inconveniences this has caused
> 
> Hi @tadahh, @kenny-buckles, have you tried these workarounds?

Hey @CristiCanizales,

Yes `Apex Setting: Perform Synchronous Compile on Deploy` is disabled 

<img width="913" alt="image" src="https://github.com/forcedotcom/salesforcedx-vscode/assets/5883817/3d45197c-f1dc-4ecf-b44a-f193b6b0d424">

Yes `Salesforcedx-vscode-core > Experimental: Enable Source Tracking For Deploy And Retrieve` is disabled

<img width="898" alt="image" src="https://github.com/forcedotcom/salesforcedx-vscode/assets/5883817/bd1d556c-ca83-46ae-a387-68137b3ad677">

I have tried everything I could find on the internet. This issue is for an existing dev sandbox we refresh every couple months.

**Comment by DoganCeliktemel** [comment]
> > Hello! Just wanted to know if you all have the `Apex Setting: Perform Synchronous Compile on Deploy` (which you can find **in your org settings**, in setup) enabled, and if so try:
> > 
> > * disable setting
> > * go to vscode and reauth to your org
> > * reload vscode
> > 
> > and see if it helps
> > Also, we want to know if the slow deployment happens to all your orgs or if it's ok with a new one
> > We're asking all these q's because disabling the setting or working with a new org has fixed it for some users. And of course disabling our vscode setting: `Salesforcedx-vscode-core > Experimental: Enable Source Tracking For Deploy And Retrieve` (and reloading vscode window after)
> > Thanks again for your patience and sorry for all the inconveniences this has caused
> 
> Hi @tadahh, @kenny-buckles, have you tried these workarounds?

Hi @CristiCanizales, thanks for the hints. However disabling the Apex Setting 'Perform Synchronous Compile on Deploy' in the org didn't change anything on the deploy times. However, I noticed that for one specific org the deploy times are especially bad. While for others they seem okay. Do you have any hint on what might be the problem for these sandboxes and how to 'clean them up' if possible to have normal deploy response times?

Thanks !

**Comment by CristiCanizales** [comment]
hi @DoganCeliktemel, it's usually an issue with source tracking, that's why we created the setting🥲
We've also seen deploy times increase after having retrieved Custom Objects from the Org Browser. Could you try deleting the .sfdx and .sf folders to clear the cache, then running "Developer: Reload Window" and re-authorizing to your org?

**Comment by tadahh** [comment]
hey @CristiCanizales,

any word on the team or any other possible solutions? We just refreshed a sandbox yesterday and I am still seeing the below issues. Deployments are taking 20+ mins. Could it be the salesforce instance server? We are on instance USA224S

<img width="1575" alt="image" src="https://github.com/forcedotcom/salesforcedx-vscode/assets/5883817/c9cf7d1d-ab1d-46ad-bb6b-dc0b66814237">

**Comment by daphne-sfdc** [comment]
@tadahh @kenny-buckles @DoganCeliktemel @mateusz-domagalski @gonardfreeman 
Following up on this issue - can we get your answers to a few questions?
1. What is your CLI version?
2. What is your VSCode version?
3. What OS are you using?
4. Are you on the latest version of our extensions?  If not, we just released v59.3.1 yesterday - please update your extensions and let us know if that fixed the issue.
5. What kind of org are you using?  (Developer, Sandbox, or Scratch?)

**Comment by daphne-sfdc** [comment]
@TejaswiniIncap Tagging you here on the original issue for slow deployment.  What kind of org are you using?  (Developer, Sandbox, or Scratch?)

**Comment by tadahh** [comment]
> @tadahh @kenny-buckles @DoganCeliktemel @mateusz-domagalski @gonardfreeman Following up on this issue - can we get your answers to a few questions?
> 
> 1. What is your CLI version?
> 2. What is your VSCode version?
> 3. What OS are you using?
> 4. Are you on the latest version of our extensions?  If not, we just released v59.3.1 yesterday - please update your extensions and let us know if that fixed the issue.
> 5. What kind of org are you using?  (Developer, Sandbox, or Scratch?)
> 
> Can you also send us your contents of your `.vscode` folder? Go to your home directory in Terminal and `cd .vscode`

1. What is your CLI version? 
`@salesforce/cli/2.15.9 darwin-x64 node-v20.9.0`

2. What is your VSCode version?
`Version: 1.84.1 (Universal)`

3. What OS are you using?
`OS: Darwin x64 22.6.0`

4. Are you on the latest version of our extensions? If not, we just released v59.3.1 yesterday - please update your extensions and let us know if that fixed the issue.
Yes, updated today

5. What kind of org are you using? (Developer, Sandbox, or Scratch?)
Sandbox Full Copy, Unlimited Edition

6. Can you also send us your contents of your .vscode folder? Go to your home directory in Terminal and cd .vscode
Folders inside .vscode - `argv.json	cli		extensions`

As of today, everything seems to be working well. The only thing I noticed, I pushed a change set with a apex class + a LWC bundle because of a prod hotfix we needed to push. After the deployment finished the UI showed an error I previously received when deploying to sandboxes when I was experiencing long deployment/conflict times

<img width="1185" alt="Screenshot 2023-10-26 at 12 59 12 PM" src="https://github.com/forcedotcom/salesforcedx-vscode/assets/5883817/475c2a44-efb0-402f-acec-1dbc1638684a">


Let me know if there is anymore info I can provide.

Thank you!

**Comment by TejaswiniIncap** [comment]
@daphne-sfdc We are currently using Salesforce Sandbox org- Unlimited Edition.
Below is the Salesforce CLI extension version that we are currently using-
![image](https://github.com/forcedotcom/salesforcedx-vscode/assets/90377637/b623b94b-4ebf-41ee-86da-a909ba45a4b7)

Below is the VS Code version that we are on, currently-
![image](https://github.com/forcedotcom/salesforcedx-vscode/assets/90377637/f9373b8b-8b54-4f49-be88-97402c9b9368)

**Comment by diyer** [comment]
Hi all, we're excited to let you know the latest version of our extensions 59.5.1 is released and we have attempted to fix this issue in our latest release. Please give the latest version a try and let us know if this addresses your concerns. We appreciate your feedback and patience! Happy exploring!

**Comment by ehez** [comment]
Hey, I updated to v59.5.1 and it was faster for a couple of days but then it started to be very slow again. Its the Conflict detection that is taking so long.

**Comment by daphne-sfdc** [comment]
@ehez Our team has a few questions for you:
1. Are you seeing slowness \*only\* in conflict detection?  Or is it also happening in deploy and retrieve operations?
2. Since you said that it was faster for a couple days then became slow again, what happens if you quit VSCode and restart it?  Does that improve conflict detection speed?
3. Can you try deploying to a new scratch org for a few days and let us know if you are seeing slowdown in that org as well after that time?

**Comment by ehez** [comment]
Hey, when I first commented, conflict detection was taking a long time. The past couple of days I disabled the setting "Detect Conflicts At Sync" in VSC and I was able to deploy much faster. I tried turning it back on today and restarted VSC. Now, deployment is taking a long time and conflict detection seems to be running much faster.

I will try deploying to a new scratch org to see what happens.

**Comment by daphne-sfdc** [comment]
@ehez Thank you for the update.  So you're seeing slowness in conflict detection \*only\* when you turn on "Detect Conflicts At Sync" in VSCode, right?  If you run the deploy right now a few times with the setting checked and a few times with the setting unchecked (alternating deploys between checked and unchecked) you're always seeing slow conflict detection only with the setting on and slow deployment only with the setting off?

**Comment by ehez** [comment]
Yeah so both conflict detection and deployment are pretty slow, especially deployment. Conflict detection takes close to a minute, deployment takes over 5 minutes.

**Comment by cadencereplies** [comment]
Latest update has brought back slow deploy times, it appears.

**Comment by kenny-buckles** [comment]
Super slow again and I am unable to run my apex tests unless I deploy by test class file each time (It won't let me click on the "run test" text to run the tests.

**Comment by JMercie** [comment]
Hello! Is anyone still looking at this issue? For instance, USA262S uses the SF CLI version @salesforce/cli/2.22.7 darwin-arm64 node-v20.10.0. It takes forever to do a deployment or to retrieve metadata. 

I'm running the command with just the -m flag to specify what I want to deploy.

**Comment by JimR-FAN** [comment]
If it helps anyone, in VSCode I reverted back to "Salesforce CLI Integration"  v58.7.1   ("Apex" extension also) 

I also have the settings  set this way:

"salesforcedx-vscode-core.experimental.enableSourceTrackingForDeployAndRetrieve": false
"salesforcedx-vscode-core.push-or-deploy-on-save.enabled": true
**"salesforcedx-vscode-core.push-or-deploy-on-save.preferDeployOnSave": true**
"salesforcedx-vscode-core.push-or-deploy-on-save.overrideConflictsOnPush": false
"salesforcedx-vscode-core.detectConflictsAtSync": false

using  @salesforce/cli/2.21.7/win32-x64 node-v20.10.0

This combination is speedy.   

From experience over the past few months, I can say that more than a few times, after an update to the extension, I've had to revert back to previous  CLI Integration extensions in order to keep things speedy. 

Thanks team for all you're doing to sort this out!

**Comment by JMercie** [comment]
> If it helps anyone, in VSCode I reverted back to "Salesforce CLI Integration" v58.7.1 ("Apex" extension also)
> 
> I also have the settings set this way:
> 
> "salesforcedx-vscode-core.experimental.enableSourceTrackingForDeployAndRetrieve": false "salesforcedx-vscode-core.push-or-deploy-on-save.enabled": true **"salesforcedx-vscode-core.push-or-deploy-on-save.preferDeployOnSave": true** "salesforcedx-vscode-core.push-or-deploy-on-save.overrideConflictsOnPush": false "salesforcedx-vscode-core.detectConflictsAtSync": false
> 
> using @salesforce/cli/2.21.7/win32-x64 node-v20.10.0
> 
> This combination is speedy.
> 
> From experience over the past few months, I can say that more than a few times, after an update to the extension, I've had to revert back to previous CLI Integration extensions in order to keep things speedy.
> 
> Thanks team for all you're doing to sort this out!

As a workaround, this helps me!

**Comment by kenny-buckles** [comment]
I downgraded my salesforce CLI, and it is still taking 30 minutes (and counting) to run apex tests. I have 3 methods in this test class. It should take a minute to run, tops.

**Comment by peternhale** [comment]
Hi @kenny-buckles, we are going to start looking at this perf problem anew and I was wondering if you could give me a synopsis of the shape of the project you are using; number of apex cls and triggers, you said you have one test class with three methods. Is that all of the test classes you have defined.

Can you also give me your platform information, which version of sfdx you might be using and the specific commands you are running.

Thanks in advance.

**Comment by kenny-buckles** [comment]
> Hi @kenny-buckles, we are going to start looking at this perf problem anew and I was wondering if you could give me a synopsis of the shape of the project you are using; number of apex cls and triggers, you said you have one test class with three methods. Is that all of the test classes you have defined.
> 
> Can you also give me your platform information, which version of sfdx you might be using and the specific commands you are running.
> 
> Thanks in advance.

I have 17 total triggers and 101 total classes (this includes managed packages).

Versions:
- @salesforce/cli/2.23.20 darwin-x64 node-v20.10.0
- Salesforce Extension Pack (Expanded) v59.13.0

The commands I am using are what is built into vscode. I am just clicking on the "Run All Tests" link in the test class:

![TcrndWly_20240116](https://github.com/forcedotcom/salesforcedx-vscode/assets/92748327/46da6d20-afa8-4d65-a4f6-330186ca3547)

**Comment by peternhale** [comment]
@kenny-buckles Thanks for the additional information. It you wouldn't mind running a simple experiment, I would like to get a notion on the time difference it takes to run the above test via `Run All Test` and then running the same test using the CLI.

Run the test using vscode and share the starting and ending timestamps from the results
Run `sf apex run test --code-coverage -d <somedir here> -n GetConfigJonTest --wait 5`

If the `sf` command times out, please increase wait value until it runs to completion.

**Comment by kenny-buckles** [comment]
@peternhale 

I have run that command. It completes, but it took about 45 minutes to complete.

**Comment by peternhale** [comment]
@kenny-buckles how does that compare to running `Run All Tests` in vscode?

Also, could you run the command again, but omit `--code-coverage`?

**Comment by kenny-buckles** [comment]
@peternhale 

Without `--code-coverage` it took 39m

**Comment by peternhale** [comment]
Thank you @kenny-buckles

**Comment by gsnortland** [comment]
It is taking 45 minutes to run a diff file against an org. I downgraded as above, to v58.7.1 and i 2.21.7.  It is still taking 45 minutes to run a diff file or a retrieve against an org, where previously it woul only take a few seconds or less than a minute.

**Comment by dima007arenko** [comment]
As for me, it is issue either at SF side or CLI. But I hope you'll dedicate all resources to resolve this issue. It is blocker for a lot of people and I don't think that fancy new features are needed when core functionality is not working properly :) 
Thank you for your hard work.

**Comment by ahenry30** [comment]
Adding this to settings.json has fixed the issue for me

> "salesforcedx-vscode-core.experimental.enableSourceTrackingForDeployAndRetrieve": false

**Comment by AnanyaJha** [comment]
Hi @dima007arenko & @ahenry30! I'm sorry to hear that you're still facing this issue. Could you please log a new issue with all your repro details? We'll have a member of the team investigate - thanks.

**Comment by peternhale** [comment]
I have opened a new issue for the slow deploy source to org, #5496 and I will be asking for some data from those that are experiencing the issue. Please visit the new issue for details.

**Comment by onelittleant** [comment]
In case this will help anyone else I would like to share our experience.  We use vscode and the deploy source to org function as a regular step in our development process.  Starting some time in late 2023 / early 2024 we started to notice that deploy source operations that we were accustomed to taking a few seconds (eg. 1 - 3 seconds) started taking longer... sometimes much much longer, like over a minute.  Early on is seemed intermittent, but it got progressively worse, to the point that developer productivity was badly impacted.  We were waiting 2 to 3 minutes to deploy a line of code and run a test that would have in the past taken less than 30 seconds for the whole deploy and test operation.  We suspect that the problem started when our org was migrated to the new salesforce cloud environment.  We read this post and many others and tried the recommended mitigations to no avail.

WE FINALLY FOUND A SOLUTION!

In our sandbox environment we went to Settings -> Apex Settings and unchecked the "Perform Synchronous Compile on Deploy" checkbox.  Since we made this change, our deploy source to org operations are back to taking a couple of seconds.  It has been life changing.  Hope this helps someone else.

**Comment by aheid173** [comment]
> WE FINALLY FOUND A SOLUTION!
> 
> In our sandbox environment we went to Settings -> Apex Settings and unchecked the "Perform Synchronous Compile on Deploy" checkbox. Since we made this change, our deploy source to org operations are back to taking a couple of seconds. It has been life changing. Hope this helps someone else.

YOU ARE AN ABSOLUTE LEGEND.  Fixed it straightaway for our whole dev team.

**Comment by jcallahan526** [comment]
> In case this will help anyone else I would like to share our experience. We use vscode and the deploy source to org function as a regular step in our development process. Starting some time in late 2023 / early 2024 we started to notice that deploy source operations that we were accustomed to taking a few seconds (eg. 1 - 3 seconds) started taking longer... sometimes much much longer, like over a minute. Early on is seemed intermittent, but it got progressively worse, to the point that developer productivity was badly impacted. We were waiting 2 to 3 minutes to deploy a line of code and run a test that would have in the past taken less than 30 seconds for the whole deploy and test operation. We suspect that the problem started when our org was migrated to the new salesforce cloud environment. We read this post and many others and tried the recommended mitigations to no avail.
> 
> WE FINALLY FOUND A SOLUTION!
> 
> In our sandbox environment we went to Settings -> Apex Settings and unchecked the "Perform Synchronous Compile on Deploy" checkbox. Since we made this change, our deploy source to org operations are back to taking a couple of seconds. It has been life changing. Hope this helps someone else.

NEED TO SECOND THIS!! THANKS!!!!
